spring-integration-jdbc/src/main/java/org/springframework/integration/jdbc/JdbcMessageHandler.java

```diff
@@ -149,57 +169,99 @@ protected void onInit() throws Exception {
 	 * Executes the update, passing the message into the {@link SqlParameterSourceFactory}.
 	 */
 	@Override
-	protected void handleMessageInternal(Message<?> message) throws Exception {
+	protected void handleMessageInternal(Message<?> message) {
 		List<? extends Map<String, Object>> keys = executeUpdateQuery(message, this.keysGenerated);
 		if (!keys.isEmpty() && logger.isDebugEnabled()) {
 			logger.debug("Generated keys: " + keys);
 		}
 	}
 
 	protected List<? extends Map<String, Object>> executeUpdateQuery(final Message<?> message, boolean keysGenerated) {
-		SqlParameterSource updateParameterSource = EmptySqlParameterSource.INSTANCE;
-		if (this.preparedStatementSetter == null) {
-			if (this.sqlParameterSourceFactory != null) {
-				updateParameterSource = this.sqlParameterSourceFactory.createParameterSource(message);
-			}
-		}
 		if (keysGenerated) {
 			if (this.preparedStatementSetter != null) {
-				return this.jdbcOperations.getJdbcOperations().execute(this.generatedKeysStatementCreator,
-						(PreparedStatementCallback<List<Map<String, Object>>>) ps -> {
-							JdbcMessageHandler.this.preparedStatementSetter.setValues(ps, message);
-							ps.executeUpdate();
-							ResultSet keys = ps.getGeneratedKeys(); // NOSONAR closed in JdbcUtils
-							if (keys != null) {
-								try {
-
-									return JdbcMessageHandler.this.generatedKeysResultSetExtractor.extractData(keys);
-								}
-								finally {
-									JdbcUtils.closeResultSet(keys);
-								}
-							}
-							return new LinkedList<Map<String, Object>>();
-						});
+				return this.jdbcOperations.getJdbcOperations()
+						.execute(this.generatedKeysStatementCreator,
+								ps -> {
+									this.preparedStatementSetter.setValues(ps, message);
+									ps.executeUpdate();
+									ResultSet keys = ps.getGeneratedKeys(); // NOSONAR closed in JdbcUtils
+									if (keys != null) {
+										try {
+
+											return this.generatedKeysResultSetExtractor.extractData(keys);
+										}
+										finally {
+											JdbcUtils.closeResultSet(keys);
+										}
+									}
+									return new LinkedList<>();
+								});
 			}
 			else {
 				KeyHolder keyHolder = new GeneratedKeyHolder();
-				this.jdbcOperations.update(this.updateSql, updateParameterSource, keyHolder);
+				this.jdbcOperations.update(this.updateSql,
+						this.sqlParameterSourceFactory.createParameterSource(message), keyHolder);
 				return keyHolder.getKeyList();
 			}
 		}
 		else {
-			int updated;
-			if (this.preparedStatementSetter != null) {
-				updated = this.jdbcOperations.getJdbcOperations().update(this.updateSql,
-						ps -> JdbcMessageHandler.this.preparedStatementSetter.setValues(ps, message));
+			if (message.getPayload() instanceof Iterable) {
+				Stream<? extends MutableMessage<?>> messageStream =
+						StreamSupport.stream(((Iterable<?>) message.getPayload()).spliterator(), false)
+								.map(payload -> new MutableMessage<>(payload, message.getHeaders()));
```

Comment:
How about...

```java
Stream<Message<?>> messageStream =
		StreamSupport.stream(((Iterable<?>) message.getPayload()).spliterator(), false)
			.map(payload -> new Message<Object>() {

				@Override
				public Object getPayload() {
					return payload;
				}

				@Override
				public MessageHeaders getHeaders() {
					return message.getHeaders();
				}
				
			});
```

...to avoid wrapping the headers in a new map for each iteration?

Reply:
Sounds like a good optimization! 👍 

Will address a bit later.

Thanks